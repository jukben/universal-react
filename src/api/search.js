// @flow

export const getRandomJoke = () =>
  fetch('https://api.chucknorris.io/jokes/random', {
    headers: new Headers({ Accept: 'application/json' }),
    method: 'GET',
  })
    .then((response) => {
      if (response.status !== 200) {
        const errorObject = {
          error: true,
          error_description: 'ChuckNorris.io API error',
        };

        throw errorObject;
      }

      return response.json();
    })
    .then(({ value: joke }) => joke)
    .catch((possibleErrorObject) => {
      if (possibleErrorObject.error) return Promise.reject(possibleErrorObject);

      const errorObject = {
        error: true,
        error_description: 'Something went wrong. Are you offline?',
      };

      throw errorObject;
    });

export default getRandomJoke;
