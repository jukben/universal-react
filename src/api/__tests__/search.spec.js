import { getRandomJoke } from '../search';
import mockSuccess from './giphy.successFetch.json';
import mockError from './giphy.errorAuth.json';

describe('API Search', () => {
  afterEach(() => {
    fetch.resetMocks();
  });

  it('should get normalized data', async () => {
    fetch.mockResponse(JSON.stringify(mockSuccess), { status: 200 });
    await expect(getRandomJoke()).resolves.toMatchSnapshot();
  });

  it('should throw error', async () => {
    fetch.mockResponse(JSON.stringify(mockError), { status: 400 });
    await expect(getRandomJoke()).rejects.toMatchSnapshot();
  });
});
