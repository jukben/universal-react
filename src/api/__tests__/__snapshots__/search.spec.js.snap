// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Search should get normalized data 1`] = `"Chuck Norris has seen it all, and done it all, so many times that he has long ago lost count."`;

exports[`API Search should throw error 1`] = `
Object {
  "error": true,
  "error_description": "ChuckNorris.io API error",
}
`;
